### PR TITLE
Dim/gendo

### PR DESCRIPTION
**Pull request by didimitrie**
Changelog: 

Backend: 
- adds feature flag for the functionality below
- adds new gendo module in server that proxies requests to the gendo api. Images (both input and genod's response) are stored as blobs (so we don't spam the db). Rate limiting on backend is not tested but it should do the trick. 

Frontend:
- makes the controls sidebar resizable - pre-requisite for the panel to actually work and allow users to enlarge their images as well as minimize panel when choosing new view
- implements new fancy gendo button + panel 
- subscriptions are live (in a rough way) so we can live display new renders as they come in

Renders are linked to each specific **version** rather than model, though they could be linked by model too if we think that's better.

TODOs in extra PR: 
- ~~helm charty things~~
- env variables
- etc. 

Renders can be downloaded as well as previewed :)

image

image

and in light mode: 

image

### CODE REVIEW

File: packages/server/modules/gendo/migrations/20240522130000_gendo.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import { Knex } from 'knex'
+import { Environment } from '@speckle/shared'
+
+// TODO: Ask whether migrations should be behind a feature flag or not??? probably... not? as we can get db in inconsistent state
+const { FF_GENDOAI_MODULE_ENABLED } = Environment.getFeatureFlags()
+const GENDO_TABLE_NAME = 'gendo_ai_renders'
+
+export async function up(knex: Knex): Promise<void> {
+  if (!FF_GENDOAI_MODULE_ENABLED) return
```
[the remark this answers is not in the document]
Comment (didimitrie) [reply packages/server/modules/gendo/migrations/20240522130000_gendo.ts:9]: Fixing!

File: packages/server/modules/gendo/migrations/20240522130000_gendo.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    table.jsonb('camera').notNullable()
+    table.text('baseImage').notNullable() // TODO: hopefull these will be blob urls and not the raw deal
+    table.text('responseImage')
+
+    table.index(['gendoGenerationId'])
+  })
+}
+
+export async function down(knex: Knex): Promise<void> {
+  if (!FF_GENDOAI_MODULE_ENABLED) return
```
[the remark this answers is not in the document]
Comment (didimitrie) [reply packages/server/modules/gendo/migrations/20240522130000_gendo.ts:39]: Ditto, fixin'

File: packages/server/modules/gendo/migrations/20240522130000_gendo.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import { Knex } from 'knex'
+import { Environment } from '@speckle/shared'
+
+// TODO: Ask whether migrations should be behind a feature flag or not??? probably... not? as we can get db in inconsistent state
+const { FF_GENDOAI_MODULE_ENABLED } = Environment.getFeatureFlags()
+const GENDO_TABLE_NAME = 'gendo_ai_renders'
```
[the remark this answers is not in the document]
Comment (didimitrie) [reply packages/server/modules/gendo/migrations/20240522130000_gendo.ts:6]: I'd love to do generalise like this, but I feel we need to roll out at least another extension before we do so :)

File: packages/server/modules/gendo/migrations/20240522130000_gendo.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    table
+      .timestamp('updatedAt', { precision: 3, useTz: true })
+      .defaultTo(knex.fn.now())
+      .notNullable()
+
+    table.text('gendoGenerationId') // Nullable as the api can fail on first call (?)
+    table.text('status').notNullable()
+    table.text('prompt').notNullable()
+    table.jsonb('camera').notNullable()
+    table.text('baseImage').notNullable() // TODO: hopefull these will be blob urls and not the raw deal
```
[the remark this answers is not in the document]
Comment (didimitrie) [reply packages/server/modules/gendo/migrations/20240522130000_gendo.ts:31]: I've clarified what they are - there was a bit of an evolution as we went. Initially base64 images - like our avatars - then proper blobIds :)

**Review by gjedlicska — APPROVED** [review]

File: utils/helm/speckle-server/templates/server/deployment.yml [hunk 5]
```diff
@@ -158,6 +158,21 @@ spec:
             value: {{ .Values.server.migration.movedTo }}
             {{- end }}
 
+          # *** Gendo render module ***
+          - name: FF_GENDOAI_MODULE_ENABLED
+            value: {{ .Values.featureFlags.gendoModuleEnabled }}
+            
+            {{- if .Values.featureFlags.gendoModuleEnabled }}
+          - name: GENDOAI_KEY
+            value: {{ .Values.server.gendoKey }}
```
Comment (gjedlicska) [inline utils/helm/speckle-server/templates/server/deployment.yml:167]: The two keys need to be secret references, they should be loaded from the `server-vars` secret

**Review by gjedlicska — CHANGES_REQUESTED** [review]

**Review by gjedlicska — APPROVED** [review]

**Pull request merged**